### PR TITLE
cmd/gitannex: Add the gitannex command

### PR DESCRIPTION
**Pull request by dmcardle**
This PR adds a new subcommand named "gitannex", aka "git-annex-remote-rclone-builtin" when invoked via a symlink.
  
This accomplishes milestone 1 from issue #7625: "minimal support for the external special remote protocol".

-----

#### What is the purpose of this change?

To add a shim program that enables git-annex to interact with rclone remotes.

#### Was the change discussed in an issue or in the forum before?

* #7625 
* https://forum.rclone.org/t/new-subcommand-for-git-annex-remote/44546

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by dmcardle** [comment]
@ncw This is ready for review!

First of all, would you be willing to accept this new command in contrib/ or would you prefer I publish as a separate project? I like it living in the rclone repo because it uses rclone as a library — and I get to abuse your CI :).

At first, I wanted to add it as a new subcommand, e.g `rclone gitannex`, but it turns out that git-annex is incapable of invoking a "special remote program" with arguments. Rather, it just executes `git-annex-remote-${REMOTE_TYPE}` by design. I considered making a wrapper shell script with the expected name, but that would not be cross-platform. Ultimately, I settled on this design, where we just have a separate binary living in contrib/.

I wouldn't ask you to include this command in releases because it would probably double the size. Also, it's only useful to a very small, technically inclined crowd who can probably figure out how to install it themselves.

Thanks!

File: contrib/gitannex/git-annex-remote-rclone-goyle/main.go
Comment (dmcardle) [inline contrib/gitannex/git-annex-remote-rclone-goyle/main.go:79]: I suppose I don't need this. It was originally included to align with rclone subcommands.

File: contrib/gitannex/git-annex-remote-rclone-goyle/main.go
Comment (dmcardle) [inline contrib/gitannex/git-annex-remote-rclone-goyle/main.go:75]: I should mention that I followed my own instructions and I'm already using the command to back up my own content :)

File: .gitignore
Comment (dmcardle) [inline .gitignore:6]: The line above begins with slash to disambiguate from the contrib/ subdirectory with the same name.

This one doesn't need to begin with a slash, I suppose.

File: contrib/gitannex/git-annex-remote-rclone-goyle/main.go
Comment (ncw) [inline contrib/gitannex/git-annex-remote-rclone-goyle/main.go:77]: That is the rclone version the subcommand was introduced.

File: contrib/gitannex/git-annex-remote-rclone-goyle/main.go
Comment (ncw) [inline contrib/gitannex/git-annex-remote-rclone-goyle/main.go:1]: Do you think this should be an rclone command so it is always built in to rclone? So something like `rclone git annexe` or something (don't really know what should be there!).

If not then this could live in its own repo quite easily? I'd be happy to host that repo under the `rclone` org.

File: contrib/gitannex/gitannex.go
Comment (ncw) [reply contrib/gitannex/gitannex.go:319]: > This should probably be a two step process: (1) transfer to a temp file with a random component to avoid collisions, (2) rename the temp file to the original key.

Rclone will do this automatically when transferring to a backend where partially uploaded files are visible (eg the local backend or FTP).

> I don't know what happens on every rclone backend if two rename operations race

Depends on the backend but likely last one wins.

**Review by ncw — COMMENTED** [review]
I've given the code a light skim :-) It looks very well written.

I think we should give some thought to whether we should make this an rclone command or not.

If we make it an rclone command then it will be built in and easier to deploy for the users (without having to download go). It will of course bloat the binary with stuff, but that is a bridge we crossed a long time ago!

If it isn't an rclone command then it should probably live in its own repository (say `rclone/gitannex`). If you like that idea I'd be happy to make that repository and give you r/w access to it.

Let me know what you think?

File: contrib/gitannex/git-annex-remote-rclone-goyle/main.go
Comment (dmcardle) [reply contrib/gitannex/git-annex-remote-rclone-goyle/main.go:1]: I thought this should be a subcommand originally, but it turns out that git-annex wants to exec `git-annex-remote-${type}` with no arguments, so there goes that idea. It seems a little silly to add a subcommand to rclone that requires a separate wrapper script/command to be useful.

> If not then this could live in its own repo quite easily? I'd be happy to host that repo under the rclone org.

Thanks for the offer to create a new repo! Unfortunately, my employer requires me to jump through more hoops if I'm creating a new project than if I'm patching an existing project. (I think even more hoops if I'm not releasing it under their GitHub org.) I'm going to research what the options are a little more and see how hard it might be to release in the rclone org.

So ultimately, I tried to thread the needle by putting this in contrib/,  satisfying both technical and bureaucratic constraints. Another advantage of this approach is that the project inherit's rclone's CI configuration and other infrastructure. I suppose the  disadvantage is that I have to bug you folks when I want to merge a PR, but it's nice to have another set of eyes :)

Comment (ncw) [reply contrib/gitannex/git-annex-remote-rclone-goyle/main.go:1]: > I thought this should be a subcommand originally, but it turns out that git-annex wants to exec `git-annex-remote-${type}` with no arguments, so there goes that idea. It seems a little silly to add a subcommand to rclone that requires a separate wrapper script/command to be useful.

Sorry, I forgot to mention the other part of the plan which would be to create a hardlink from `git-annex-remote-xxxx` to rclone then get rclone to start the correct subcommand. We already do this for [mount helpers](https://rclone.org/commands/rclone_mount/#rclone-as-unix-mount-helper) so this is easy to do.

> > If not then this could live in its own repo quite easily? I'd be happy to host that repo under the rclone org.
> 
> Thanks for the offer to create a new repo! Unfortunately, my employer requires me to jump through more hoops if I'm creating a new project than if I'm patching an existing project. (I think even more hoops if I'm not releasing it under their GitHub org.) I'm going to research what the options are a little more and see how hard it might be to release in the rclone org.

I could create the project under the rclone org then invite you to collaborate - perhaps that would satisfy your employers?

> So ultimately, I tried to thread the needle by putting this in contrib/, satisfying both technical and bureaucratic constraints. Another advantage of this approach is that the project inherit's rclone's CI configuration and other infrastructure. 

The CI is a lot of work to set up, I agree! If it is part of rclone it could also participate in the integration tests.

> I suppose the disadvantage is that I have to bug you folks when I want to merge a PR, but it's nice to have another set of eyes :)

:-)

Comment (dmcardle) [reply contrib/gitannex/git-annex-remote-rclone-goyle/main.go:1]: > Sorry, I forgot to mention the other part of the plan which would be to create a hardlink from `git-annex-remote-xxxx` to rclone then get rclone to start the correct subcommand. We already do this for [mount helpers](https://rclone.org/commands/rclone_mount/#rclone-as-unix-mount-helper) so this is easy to do.

Oh, cool! I didn't realize this was an option!

I think making this a subcommand is the best option, then. This also means I don't need to jump through approval hoops.

Thanks for the pointer to the mount helpers. IIUC, it looks like fs/mount_helpers.go transforms args in its `init()` *before*  rclone's `Main()` function runs. I can probably pull the same trick in the new subcommand's `init()`.

Comment (ncw) [reply contrib/gitannex/git-annex-remote-rclone-goyle/main.go:1]: > I think making this a subcommand is the best option, then. This also means I don't need to jump through approval hoops.

Great

> Thanks for the pointer to the mount helpers. IIUC, it looks like fs/mount_helpers.go transforms args in its init() before rclone's Main() function runs. I can probably pull the same trick in the new subcommand's init().

I can't remember exactly how that works, but I'm sure you'll figure it out!

File: .gitignore
Comment (dmcardle) [reply .gitignore:6]: Leaving this as is, it doesn't seem super important.

**Comment by dmcardle** [comment]
Hey @ncw, this is ready for another look when you get a chance.

I factored it back into a subcommand (`rclone gitannex`). It now recognizes when rclone is invoked via a symlink (or whatever) named `git-annex-remote-rclone-goyle`.

Thanks!

**Comment by dmcardle** [comment]
Just made a few little improvements, like more consistent use of the `messageParser` type and more test coverage.

**Comment by joeyh** [comment]
WRT "rclone gitannex" vs "git-annex-remote-rclone-goyle" I would be open to considering doing something in git-annex to support the former without needing the latter. If that would be helpful enough to bother with. It might be as simple as a hardcoded special case.

**Comment by dmcardle** [comment]
> WRT "rclone gitannex" vs "git-annex-remote-rclone-goyle" I would be open to considering doing something in git-annex to support the former without needing the latter. If that would be helpful enough to bother with. It might be as simple as a hardcoded special case.

That would be really useful, @joeyh! Although I would probably keep the symlink detection logic in this PR in order to support current releases of git-annex. I'd keep it around until a git-annex release with the new feature trickled down to Debian stable.

My taste would be to implement a more generic mechanism rather than adding a special case for `rclone gitannex`. What if `externaltype` could be repeated, so that `git annex initremote MyRemote type=external externaltype=rclone externaltype=gitannex ...` would cause git-annex to exec `rclone` with the additional `gitannex` arg?

File: cmd/gitannex/run-git-annex-testremote.sh
Comment (yarikoptic) [inline cmd/gitannex/run-git-annex-testremote.sh:10]: best to avoid any assumptions or otherwise would require per location where you would intend to run this test to set it up somehow etc. What about the approach we have in git-annex-remote-rclone to use temporary HOME with all needed setup? checkout 

https://github.com/git-annex-remote-rclone/git-annex-remote-rclone/blob/master/tests/all-in-one.sh#L22

which might also give you ideas on additional (it is never enough!!) testing etc.

File: cmd/gitannex/gitannex.go
Comment (yarikoptic) [inline cmd/gitannex/gitannex.go:43]: `goyle` is cute but may be more boring `native` or `builtin` thus to signal it's intimate relationship to rclone itself?

Comment (ncw) [reply cmd/gitannex/gitannex.go:43]: I have to say I don't understand what "goyle" or "garrgoyle" means at all! Do we need to expose it to users?

File: cmd/gitannex/gitannex.md
Comment (ncw) [inline cmd/gitannex/gitannex.md:15]: Could the type just be `rclone` rather than `rclone-goyle`? As someone who knows very little about the protocols involved, the word `goyle` means nothing to me - is it important?

**Review by ncw — COMMENTED** [review]
I think this is looking very good and ready to merge.

I'd like to make sure we get the naming right - that is the only thing that is difficult to change afterwards - your opinions on that would be great.

File: cmd/gitannex/gitannex.go
Comment (dmcardle) [reply cmd/gitannex/gitannex.go:43]: So I was trying to be clever with the name, but I've gotten pretty consistent feedback that the name is confusing!

We need to expose *some* name to users since they'll need to install a symlink somewhere (unless/until @joeyh adds a special-case to git-annex!).

I'll rebrand it as git-annex-remote-rclone-goyle and get rid of the "garrgoyle" name. Thanks, folks!

File: cmd/gitannex/gitannex.md
Comment (dmcardle) [reply cmd/gitannex/gitannex.md:15]: So if the name were just "rclone", then we would have a name collision with <https://github.com/git-annex-remote-rclone/git-annex-remote-rclone>. I'm working on changing the type to "rclone-builtin" right now, which seems reasonably self explanatory.

Back when I was thinking this would be a standalone command, rather than baked into rclone, I picked a globally-unique suffix that also made a semi-pun (git-annex-remote-rclone-goyle ==> "garrgoyle").

File: cmd/gitannex/run-git-annex-testremote.sh
Comment (dmcardle) [reply cmd/gitannex/run-git-annex-testremote.sh]: @yarikoptic Very clever! In the interest of unblocking the PR, I'll punt on improving the test script for now, but this is the very next thing on my TODO list for the project. I really appreciate you taking a look at the PR.

Comment (yarikoptic) [reply cmd/gitannex/run-git-annex-testremote.sh]: Traveling ATM, can't promise speedy review over weekend, but would be happy to help in general.

**Comment by dmcardle** [comment]
> I think this is looking very good and ready to merge.
> 
> I'd like to make sure we get the naming right - that is the only thing that is difficult to change afterwards - your opinions on that would be great.

Hey @ncw, just got rid of the confusing name. Now it's `git-annex-remote-rclone-builtin` when invoked via symlink.  I think this is ready to go!

Also, thank you to @joeyh and @yarikoptic for taking a look!

**Review by ncw — APPROVED** [review]
Thank you for this great feature :-)

I'll merge it now.

**Comment by joeyh** [comment]
@dmcardle good point that changes in git-annex would take quite a while to trickle down to eg debian stable. Of course, so will the new rclone.

I've opened a todo on git-annex https://git-annex.branchable.com/todo/external_special_remotes_not_using_git-annex-remote_in_name/

**Comment by yarikoptic** [comment]
FWIW a shameless plug which might come handy:

- we provide non-official git-annex builds daily for the most of the OSes from https://github.com/datalad/git-annex/ 
- https://github.com/datalad/datalad-installer also has `rclone` among its components with a good range of "methods" of installation: https://github.com/datalad/datalad-installer/?tab=readme-ov-file#rclone  .

Overall -- might be easy to try the tandem as soon as we get `rclone` with it out.

**Comment by dmcardle** [comment]
Neat! Thank you for the context, @yarikoptic!

File: cmd/gitannex/gitannex.md
Comment (yarikoptic) [inline cmd/gitannex/gitannex.md:10]: I need to learn this dark magic (if it works ;-) yet to get a chance to try, but I have concerns ).

But ideally IMHO `rclone` itself should just install a shim itself in the PATH. The trickery that it would be different for *nix (shell script is enough) and Windows (needs .bat or .exe?)

Comment (dmcardle) [reply cmd/gitannex/gitannex.md:10]: Please do give it a try! I'm happy to make changes if there's something broken or risky.

The dark magic happens in gitannex.go's `maybeTransformArgs()`, which is called by `init()`. Basically, if `argv[0]` looks like "git-annex-remote-rclone-builtin", we insert the `gitannex` subcommand in the second position. There's actually already precedent in rclone for rewriting arguments like this: <https://rclone.org/commands/rclone_mount/#rclone-as-unix-mount-helper>.

It's decidedly a hack, but it means we don't have to write any platform-specific wrapper scripts. I think we can do the rough equivalent of `ln -s` on Windows with `mklink`, but I don't have a Windows machine handy to try it out. 

I see a few paths forward that enable us to remove the symlink step from the setup instructions:

* If git-annex adds a special case for this remote, i.e. @joeyh resolves this [todo](https://git-annex.branchable.com/todo/external_special_remotes_not_using_git-annex-remote_in_name/), users could just set the `type` or `externaltype` to whatever git-annex expects. No need for a symlink.
* We could write wrapper scripts for every supported platform. Honestly, this is probably not that hard, but I'm generally wary of subtleties introduced by shell scripts.
* We could write wrapper scripts for every supported platform and add a sub-subcommand (something like `rclone gitannex installwrapper`) that installs the platform-specific wrapper somewhere. That location could be alongside the `rclone` binary, but that makes me nervous. It could be a user-selected bin directory, but then the user still has to manage their PATH.
* We could write wrapper scripts for every supported platform and ship them in rclone releases OR we could even ship relative symlinks in the releases, I suppose. This solves the package manager issue, but I'm not sure how project maintainers would feel about adding files to the release.

Comment (yarikoptic) [reply cmd/gitannex/gitannex.md:10]: sorry -- my bad -- I was just too tired I guess. Instead of `realpath rclone` I saw `rclone gitannex` ;-)

Thanks for the analysis!  yeah - we need @ncw's input on this.  I guess such a wrapper could be done directly in `go` and just come with `rclone` package "natively" to not mess with "platform specific solutions" (no symlinks on windows really AFAIK).

Comment (ncw) [reply cmd/gitannex/gitannex.md:10]: > I see a few paths forward that enable us to remove the symlink step from the setup instructions:

You could get rclone to create the symlink on the first writable element in PATH it finds...

I'd avoid scripting if possible - Go is marvelously consistent between platforms whereas shell/powershell isn't!

If we can build the logic into rclone then that makes it a one binary deploy which is what we aim for.

**Comment by joeyh** [comment]
@dmcardle I've implemented the rclone special remote in git-annex, so no symlink is needed and config looks like:

git-remote foo type=rclone encryption=none rcloneremotename=foo rcloneprefix=bar

Could you update the rclone documentation to reflect this?

**Comment by dmcardle** [comment]
> Could you update the rclone documentation to reflect this?

Will do!

@joeyh, I have to admit I don't really understand the command you posted — is it supposed to be `git annex initremote` rather than `git-remote`?

FYI, I'm working on making this compatible with the various layout modes defined by git-annex-remote-rclone in #7754. Can you advise on how to switch an existing remote from `externaltype=rclone` to `type=rclone`? (Or is that a bad idea?) I'm hoping that users can transition to rclone's built-in remote without needing to copy their data.

**Comment by joeyh** [comment]
Sorry, that was in fact supposed to start with git-annex initremote.

**Pull request merged**
